base/twiceprecision.jl

```diff
@@ -193,27 +193,40 @@ roundoff error).  If `step` has an exact rational representation
 where `nb` is the number of trailing zero bits of `step.hi`.  For
 ranges, you can set `nb = ceil(Int, log2(len-1))`.
 """
-struct TwicePrecision{T}
+struct TwicePrecision{T<:Number}
     hi::T    # most significant bits
     lo::T    # least significant bits
 end
 
-TwicePrecision{T}(x::T) where {T} = TwicePrecision{T}(x, zero(T))
+(::Type{T})(x::TwicePrecision{T}) where {T<:Number} = x.hi
 
-function TwicePrecision{T}(x) where {T}
-    xT = convert(T, x)
+# `twiceprecision_to_number` and `to_twiceprecision` are implementation
+# detail of the constructors. Their purpose is code reuse with
+# `complex.jl`.
+
+twiceprecision_to_number(::Type{T}, x::TwicePrecision) where {T<:Number} =
```

Comment:
Isn't this semantically `convert(T, ::TwicePrecision)`? Why not overload that? After all, `convert` has precedent of being lossy, at least on `AbstractFloat`.